### PR TITLE
akka-http-testkit websocket support (#17914) + WS subprotocols in WebsocketDirectives (#17726)

### LINKED ISSUE #17726: Add API for Websocket subprotocol negotiation to WebsocketDirectives
See https://github.com/akka/akka/issues/17609#issuecomment-111828935

/cc @danielwegener @sirthias

### PR DESCRIPTION
**Pull request by jrudolph**
Will fix #17726 and #17914.

/cc @sirthias

### CODE REVIEW

**Comment by jrudolph** [comment]
Added documentation.

**Comment by jrudolph** [comment]
This is ready for review from my side.

**Comment by jrudolph** [comment]
Error lines in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/3497/:

```
 12651 Failed: Total 528, Failed 1, Errors 0, Passed 527, Ignored 467
 12652 Failed tests:
 12653 	akka.stream.tck.SynchronousFilePublisherTest
 16032 (akka-stream-tck-experimental/test:test) sbt.TestsFailedException: Tests unsuccessful
 16033 Total time: 889 s, completed Aug 24, 2015 2:03:17 PM
```

**Comment by jrudolph** [comment]
Test failure seems to be unrelated.

File: akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/WebsocketExampleSpec.scala [hunk 1]
```diff
@@ -27,13 +29,16 @@ class WebsocketExampleSpec extends WordSpec with Matchers {
...
           // we match but don't actually consume the text message here,
           // rather we simply stream it back as the tail of the response
           // this means we might start sending the response even before the
           // end of the incoming message has been received
-          case tm: TextMessage ⇒ TextMessage(Source.single("Hello ") ++ tm.textStream)
-          // ignore binary messages
+          case tm: TextMessage ⇒ TextMessage(Source.single("Hello ") ++ tm.textStream) :: Nil
+          case bm: BinaryMessage =>
+            // ignore binary messages but drain content to avoid the stream being clogged
+            bm.dataStream.runWith(Sink.ignore)
```
Comment (drewhk) [inline akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/WebsocketExampleSpec.scala]: nice catch.

File: akka-http-testkit/src/main/scala/akka/http/scaladsl/testkit/WSProbe.scala [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  def inProbe: TestSubscriber.Probe[Message]
+
+  /**
+   * The underlying probe for the ingoing side of this probe. Can be used if the methods
+   * on WSProbe don't allow fine enough control over the message flow.
+   */
+  def outProbe: TestPublisher.Probe[Message]
+}
+
+object WSProbe {
```
Comment (drewhk) [inline akka-http-testkit/src/main/scala/akka/http/scaladsl/testkit/WSProbe.scala]: Nice!

File: akka-http-testkit/src/main/scala/akka/http/scaladsl/testkit/WSTestRequestBuilding.scala [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import akka.http.impl.engine.ws.InternalCustomHeader
+import akka.http.scaladsl.model.headers.{ UpgradeProtocol, Upgrade, `Sec-WebSocket-Protocol` }
+import akka.http.scaladsl.model.{ StatusCodes, HttpResponse, HttpRequest, Uri }
+import akka.http.scaladsl.model.ws.{ UpgradeToWebsocket, Message }
+import akka.stream.scaladsl.Flow
+
+import scala.collection.immutable
+
+trait WSTestRequestBuilding { self: RouteTest ⇒
+  def WS(uri: Uri, clientSideHandler: Flow[Message, Message, Any], subprotocols: Seq[String] = Nil)(): HttpRequest =
```
Comment (drewhk) [inline akka-http-testkit/src/main/scala/akka/http/scaladsl/testkit/WSTestRequestBuilding.scala]: Nice, too!

**Comment by drewhk** [comment]
LGTM, very nice test tools!

**Comment by drewhk** [comment]
PLS BUILD

**Comment by jrudolph** [comment]
Error lines in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/3509/:

```
 16014 Failed: Total 648, Failed 1, Errors 0, Passed 647, Pending 7
 16015 Failed tests:
 16016 	akka.http.scaladsl.server.directives.FileAndResourceDirectivesSpec
 16035 (akka-http-tests-experimental/test:test) sbt.TestsFailedException: Tests unsuccessful
 16036 Total time: 833 s, completed Aug 25, 2015 2:17:36 PM
```

**Comment by jrudolph** [comment]
See #18318 for the reason of the test failure.

**Comment by jrudolph** [comment]
Rebased.

**Comment by drewhk** [comment]
LGTM. Shoud I merge or do you plan to add anything?

**Comment by jrudolph** [comment]
You can merge this one has been finished.

**Pull request merged**